cmd/hook.go

```diff
@@ -460,3 +473,261 @@ func pushOptions() map[string]string {
+			if err != nil {
+				fail("Internal Server Error", "Pkt-Line format is wrong :%v", err)
+			}
+			if rs.Type == pktLineTypeFlush {
+				break
+			}
+
+			kv := strings.SplitN(rs.Data, "=", 2)
+			if len(kv) == 2 {
+				hookOptions.GitPushOptions[kv[0]] = kv[1]
+			}
+		}
+	}
+
+	// run hook
+	resp, err := private.HookProcReceive(repoUser, repoName, hookOptions)
+	if err != nil {
+		fail("Internal Server Error", "run proc-receive hook failed :%v", err)
+	}
+
+	// 3 response result to service.
+	// # OK, but has an alternate reference.  The alternate reference name
+	// # and other status can be given in option directives.
+	// H: PKT-LINE(ok <ref>)
+	// H: PKT-LINE(option refname <refname>)
+	// H: PKT-LINE(option old-oid <old-oid>)
+	// H: PKT-LINE(option new-oid <new-oid>)
+	// H: PKT-LINE(option forced-update)
+	// H: ... ...
+	// H: flush-pkt
+	for _, rs := range resp.Results {
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("ok "+rs.OrignRef))
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("option refname "+rs.Ref))
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("option old-oid "+rs.OldOID))
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+		err = writePktLine(os.Stdout, pktLineTypeData, []byte("option new-oid "+rs.NewOID))
+		if err != nil {
+			fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+		}
+	}
+	err = writePktLine(os.Stdout, pktLineTypeFlush, nil)
+	if err != nil {
+		fail("Internal Server Error", "Pkt-Line response failed: %v", err)
+	}
+	return nil
+}
+
+// git PKT-Line api
+// pktLineType message type of pkt-line
+type pktLineType int64
+
+const (
+	// UnKnow type
+	pktLineTypeUnknow pktLineType = 0
+	// flush-pkt "0000"
+	pktLineTypeFlush pktLineType = iota
+	// data line
+	pktLineTypeData
+)
+
+// gitPktLine pkt-line api
+type gitPktLine struct {
+	Type   pktLineType
+	Length int64
+	Data   string
+}
+
+func readPktLine(in *bufio.Reader) (r *gitPktLine, err error) {
+	// read prefix
+	lengthBytes := make([]byte, 4)
+	for i := 0; i < 4; i++ {
+		lengthBytes[i], err = in.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+	}
+	r = new(gitPktLine)
+	r.Length, err = strconv.ParseInt(string(lengthBytes), 16, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.Length == 0 {
+		r.Type = pktLineTypeFlush
+		return r, nil
+	}
+
+	if r.Length <= 4 || r.Length > 65520 {
+		r.Type = pktLineTypeUnknow
+		return r, nil
+	}
+
+	tmp := make([]byte, r.Length-4)
+	for i := range tmp {
+		tmp[i], err = in.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	r.Type = pktLineTypeData
+	r.Data = string(tmp)
+
+	return r, nil
+}
+
+func writePktLine(out io.Writer, typ pktLineType, data []byte) error {
+	if typ == pktLineTypeFlush {
```

Comment:
You can define two different functions, one write pkt-line, and the other write flush packet.